website/src/pages/api/auth/[...nextauth].ts

```diff
@@ -94,20 +99,44 @@ const authOptions: AuthOptions = {
       session.user.role = token.role;
       session.user.isNew = token.isNew;
       session.user.name = token.name;
+      session.user.tosAcceptanceDate = token.tosAcceptanceDate;
       return session;
     },
     /**
      * When creating a token, fetch the user's role and inject it in the token.
      * This let's use forward the role to the session object.
      */
     async jwt({ token }) {
-      const { isNew, name, role } = await prisma.user.findUnique({
+      const { isNew, name, role, accounts, id } = await prisma.user.findUnique({
         where: { id: token.sub },
-        select: { name: true, role: true, isNew: true },
+        select: { name: true, role: true, isNew: true, accounts: true, id: true },
       });
+
+      const user: BackendUserCore = {
+        id,
+        display_name: name,
+        auth_method: accounts.length > 0 ? accounts[0].provider : "local",
+      };
+      const oasstApiClient = createApiClientFromUser(user);
+      let tosAcceptanceDate;
+      try {
+        /**
+         * when first creating a new user, we have a race condition here,
+         * where the python backend still does not know about the user
+         * in the frontend, when the user accepts the tos, we do a full refresh
+         * which means this function will be called again.
+         */
+        tosAcceptanceDate = await oasstApiClient.fetch_tos_acceptance(user);
+      } catch (err) {
+        if (err.httpStatusCode !== 404) {
+          throw err;
+        }
+      }
+
```

Comment:
@fozziethebeat do you have any idea how to overcome this? or am I misinterpreting the situation?

Reply:
from @andreaskoepf  on discord:

> I don't think that is already happening .. I added an endpoint for create-or-update user only very recently ... I fear the old code only indirectly notified the backend by sending a new user for the next-task query.

So this should be on our todo list

> Create or Update can be done by POST to  /api/v1/frontend_users/ ...
It returns the created/updated user info ...
(fields that are null are ignored .. eg. any of enabled, show_on_leaderboard, tos_acceptance, notes ) 


I have updated the comments to reflect this.